modules/dnn/src/layers/fully_connected_layer.cpp

```diff
@@ -415,6 +422,80 @@ class FullyConnectedLayerImpl CV_FINAL : public InnerProductLayer
         }
     }
 
+#ifdef HAVE_CUDA
+    void forwardCUDA(
+        std::vector<cv::Ptr<BackendWrapper>>& inputs,
+        std::vector<cv::Ptr<BackendWrapper>>& outputs,
+        csl::Workspace& workspace
+    )
+    {
+        CV_UNUSED(workspace);
+
+        for (std::size_t i = 0; i < inputs.size(); i++)
+        {
+            auto input_wrapper = inputs[i].dynamicCast<CUDABackendWrapperFP32>();
+            auto output_wrapper = outputs[i].dynamicCast<CUDABackendWrapperFP32>();
+
+            auto input = input_wrapper->getView();
+            auto output = output_wrapper->getSpan();
+
+            auto actual_dims = input_wrapper->getShape().size();
+            CV_Assert(csl::tensor_utils::get_effective_rank(input) <= actual_dims);
+
+            auto extra_dims = input.rank - actual_dims;
+            auto flatten_start_axis = clamp(axis, actual_dims) + extra_dims;
+
+            std::size_t batch_size = 1;
+            for (int j = 0; j < flatten_start_axis; j++)
+                batch_size *= input.get_axis_size(j);
+
+            auto input_size = input.size() / batch_size;
+            auto output_size = output.size() / batch_size;
+
+            CV_Assert(input_size == weightsTensor.get_axis_size(-1));
+            CV_Assert(output_size == weightsTensor.get_axis_size(-2));
+
+            for (std::size_t j = 0; j < batch_size; j++)
+            {
```

Comment:
I don't know if there is a better way to perform batch matrix operations using cuBLAS. Each batch item would require two kernel launches.

Reply:
I don’t think you can do the bias addition with cublas as I recall. However, you most definitely should be doing this using one matrix matrix multiplication followed by one bias addition. This code here looks like it kicks off a whole bunch of matrix vector multiplies. 

More generally, there shouldn’t be any loops over batch size on the cpu. The whole batch must always be processed in one kernel launch to obtain good GPU utilization.

Reply:
I will use `cublas<t>gemmStridedBatched` for multiplication. I'll write my own kernel for batched matrix addition.

There was a negligible performance difference between the cublas geam and my naive matrix addition kernel in the prototype.

I could also invoke cublas geam for each batch item inside a kernel but I think it will be as bad as launching multiple kernels from the host. I'll have to check.

Reply:
This will push the minimum CUDA toolkit version required to 8.0.

Reply:
FC (A = 500 x 800, B = 800 x 1) with N = 10,000:
batched sgemm: 171ms
cpu loop: 370ms

FC (A = 10 x 500, B = 500 x 1) with N = 10,000:
batched sgemm: 2ms
cpu loop: 87ms

Reply:
You need to compare to `cublasSgemm` as well.  My recollection is that `cublasSgemmStridedBatched` and it's other `batched` versions are awful.  You don't need to use any special batched version of `gemm` to do a fully connected layer, just a normal matrix matrix multiply can do it all. For example, this is the forward function in dlib's fully connected layer: https://github.com/davisking/dlib/blob/master/dlib/dnn/layers.h#L1619.  Note there is just one `gemm` call that internally launches one cuda kernel.

Don't launch kernel's from inside other kernels. 

Anyway, think about a tensor as a matrix that is `num_batches` rows and `everything else in the tensor` columns.  You can then do the fully connected layer with just a normal matrix matrix multiply. That's the fastest way to do this, unless one of the `gemmStridedBatched` methods has gotten faster in recent CUDA implementations.